crates/lang_srv/src/registry.rs

```diff
@@ -1,75 +1,198 @@
+                            info: old_doc.info.clone(),
+                            latest_document: old_doc.latest_document.clone(),
+                            last_good_document: document,
+                        };
+                    }
+                } else if document.type_checked() {
+                    *old_doc = DocumentPair::new(document.clone(), document);
+                } else {
+                    debug!(
+                        "Document typechecking failed at version {:?}, not updating last_good_document",
+                        &document.doc_info.version
+                    );
+                    *old_doc = DocumentPair::new(document, old_doc.last_good_document.clone());
                 }
             }
-            DocumentChange::Closed(_url) => {
-                // Do nothing.
+            None => {
+                documents.insert(url.clone(), DocumentPair::new(document.clone(), document));
+            }
+        }
+    }
+
+    pub async fn apply_changes<'a>(&self, analysed_docs: Vec<AnalyzedDocument>, updating_url: Url) {
+        let mut documents = self.documents.lock().await;
+        debug!(
+            "finised doc analysis for doc: {:?}",
+            updating_url.to_string()
+        );
+
+        for document in analysed_docs {
+            let document = Arc::new(document);
+            //Write the newly analysed document into the partial document that any request requiring the latest document will be waiting on
+            Registry::update_document(&mut documents, document, &updating_url);
+        }
+    }
+
+    pub async fn apply_doc_info_changes(&self, url: Url, info: DocInfo) {
+        let mut documents_lock = self.documents.lock().await;
+        let doc = documents_lock.get_mut(&url);
+        match doc {
+            Some(a) => {
+                debug!(
+                    "set the docInfo for {:?} to version:{:?}",
+                    url.as_str(),
+                    info.version
+                );
+                *a = DocumentPair {
+                    info,
+                    last_good_document: a.last_good_document.clone(),
+                    latest_document: OnceLock::new(),
+                };
             }
+            None => debug!("no existing docinfo for {:?} ", url.as_str()),
         }
     }
 
-    fn document_by_url(&mut self, url: &Url) -> Option<&mut AnalyzedDocument> {
-        self.documents.get_mut(url)
+    async fn document_info_by_url(&self, url: &Url) -> Option<DocInfo> {
+        self.documents.lock().await.get(url).map(|a| a.info.clone())
     }
 
-    pub fn diagnostics(&mut self, url: &Url) -> Vec<Diagnostic> {
-        let Some(document) = self.document_by_url(url) else {
+    ///Tries to get the latest document from analysis.
+    ///Gives up and returns none after 5 seconds.
+    async fn latest_document_by_url(&self, url: &Url) -> Option<Arc<AnalyzedDocument>> {
+        let duration = std::time::Duration::from_secs(5);
```

Comment:
Let's make this configurable with an env var and default to five seconds. I would make it ROCLS_TIMEOUT_MS and default to 5000. Reason being I want to test this.